procrastinate/blueprints.py

```diff
@@ -0,0 +1,54 @@
+import functools
+from typing import TYPE_CHECKING, Any, Callable, List, Optional
+
+from procrastinate import jobs, retry
+
+if TYPE_CHECKING:
+    from procrastinate import app, tasks
+
+
+class BluePrint:
+    def __init__(self):
+        self.tasks = {}
+
+    def _register_task(self, task: "tasks.Task") -> None:
+        self.tasks[task.name] = task
+
+    def register(self, app: "app.App") -> None:
+        for task in self.tasks.values():
+            app._register(task)
+
+    def task(
```

Comment:
I'd be tempted to have something break if the signature for this method deviates from the one in `app.task`. This could be a test that compare signatures from `inspect` (https://docs.python.org/3/library/inspect.html) but it could also be an `abc` (https://docs.python.org/3/library/abc.html).

Though I think my prefered way would be if this could be a protocol (using `runtime_checkable`) https://mypy.readthedocs.io/en/stable/protocols.html#using-isinstance-with-protocols.

I'm not even sure we need to expose this protocol outside of tests.

Reply:
I did wonder about how to keep these two signatures inline.  

As far as I understand `abc.abstractmethod` only checks that the method exists, and does not make any assertions about the arguments.

Both inspect and protocols are type assertions.  I guess the question is whether you want any error to occur during `pytest` or `mypy`.

Reply:
I'd say both would work. We could have:
```python
class TaskRegisterer(typing.Protocol):
	...

class App(TaskRegisterer):
	...
```
and have it fail on mypy, or we could have
```python
@runtime_checkable
class TaskRegisterer(typing.Protocol):
	...

class App:
	...

# tests
@pytest.mark.parametrize("cls", [App])
def check_task_registering_api(cls):
	assert isinstance(cls, TaskRegisterer)
```

Which one do you like better ? (I think I trust pytest a bit more).

Reply:
From https://docs.python.org/3/library/typing.html#typing.runtime_checkable

> Note runtime_checkable() will check only the presence of the required methods, not their type signatures! For example, builtins.complex implements __float__(), therefore it passes an issubclass() check against SupportsFloat. However, the complex.__float__ method exists only to raise a TypeError with a more informative message.

As I read it, this will still not check the args, much as the `abc` does not.

I think using `Protocol` and relying on `mypy` may be the lowest cost check/test that can be implemented.

That said, I'm not that against writing a simple test which uses `inspect` and asserts the two functions have the same signature.  

I'll try and write the test and add the protocol to see if I run into any issues.